### PR TITLE
Add detachedTripletStep to phase1 tracking sequence and reorder iterations

### PR DESCRIPTION
**Pull request by makortel**
As discussed in May 9 TRK POG meeting https://indico.cern.ch/event/527844/contributions/2161117/attachments/1269921/1881271/slides_phase1_20160509.pdf
this PR adds detachedTripletStep (only combinations of layers 2-3-4, and from BPix only 1-2-3), and reorders the iterations as `initialStep, lowPtQuadStep, highPtTripletStep, lowPtTripletStep, detachedQuadStep, detachedTripletStep).

Please find below MultiTrackValidator plots with 1000 TTbar+PU35 events showing the effect of this PR on top of 8_1_0_pre5+#14545 (strip alignment as in run2_asymptotic GT)
https://mkortela.web.cern.ch/mkortela/tracking/validation/CMSSW_8_1_0_pre5_detachedTriplet/

Tested in CMSSW_8_1_X_2016-05-25-2300, no changes expected in non-2017 workflows. In 2017 workflows efficiency increases in pT 0.3-2 GeV and r 2-8 cm, and also fake and duplicate rates increase (all caused by the addition of detachedTripletStep, fake rate increase mitigated a bit by iteration reordering). The timing should stay pretty much the same, thanks to more optimal order of iterations.

@rovere @VinInn

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild , please test

File: RecoTracker/IterativeTracking/python/ElectronSeeds_cff.py [hunk 1]
```diff
@@ -16,8 +16,8 @@
     oldClusterRemovalInfo = cms.InputTag("initialStepSeedClusterMask")
 )
 eras.trackingPhase1PU70.toModify(pixelPairStepSeedClusterMask, oldClusterRemovalInfo = "highPtTripletStepSeedClusterMask")
-# This is a pure guess to use lowPtTripletStep for phase1 here instead of the pixelPair in Run2 configuration
-lowPtTripletStepSeedClusterMask = seedClusterRemover.clone(
+# This is a pure guess to use detachedTripletStep for phase1 here instead of the pixelPair in Run2 configuration
+detachedTripletStepSeedClusterMask = seedClusterRemover.clone(
```
Comment (makortel) [inline RecoTracker/IterativeTracking/python/ElectronSeeds_cff.py]: I'm not really sure about the necessity of the changes in this file (last commit, 3b96736) as they seem to have no effect in gsfElectrons (within 1000 ttbar+PU events), but this way the cluster masking is more consistent with run2 tracking (using last pixel-only iteration).

Comment (VinInn) [inline RecoTracker/IterativeTracking/python/ElectronSeeds_cff.py]: Once we have a baseline in IB or release EGAMMA should start to have a look to all this...

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/14643	3b96736
- configuration changes are in line with the description; affect only 2017 workflows
- jenkins tests pass and comparisons with baseline show differences in 2017 workflows: most significant changes are following iteration reordering
- tested locally in CMSSW_8_1_X_2016-05-24-1100 on top of #14621 48c74d0 and #14544  5b7e864. The summary below is mainly for 10224.0 (ttbar PU35 100 events)

- from significant product changes at the product level there is an increasse of a few % in tracks, which propagates to ~1% increase in PF candidates; there is also about 20% increase in Kshort yields
```
    653.1 ->       789.8        137     18.9   0.01     recoConversions_uncleanedOnlyAllConversions__RECO.
    260.5 ->       309.4         49     17.2   0.00     recoVertexCompositeCandidates_generalV0Candidates_Kshort_RECO.
   1914.6 ->      2127.6        213     10.5   0.01     recoConversions_allConversions__RECO.
  13606.7 ->     13229.5       -377     -2.8  -0.02     recoPFTaus_combinatoricRecoTausBoosted__RECO.
 115628.0 ->    119818.0       4190      3.6   0.18     recoTracks_generalTracks__RECO.
  59519.5 ->     60363.3        844      1.4   0.04     recoPFBlocks_particleFlowBlock__RECO.
  75415.3 ->     76230.9        816      1.1   0.03     recoPFCandidates_particleFlow__RECO.
```
- timing has an overall increase:
    - Looking just at \*Step\* modules (tracking), there is a small decrease: 10926.9 ms/evt -> 10913.2 ms/evt
    - Looking at the rest, there is a relative 2.6% increase: 8027 ms/evt => 8240 ms/evt. This is mostly concentrated in the following modules. (I will try to remove the nonsense clones of particleFlowDisplacedVertex, so that the increase is less significant overall)
```
   +0.187823      +0.03%       194.09 ms/ev ->       234.33 ms/ev particleFlowDisplacedVertex
   +0.186052      +0.03%       193.24 ms/ev ->       232.88 ms/ev particleFlowDisplacedVertexNoHF
   +0.185901      +0.03%       193.22 ms/ev ->       232.82 ms/ev particleFlowDisplacedVertexPuppi
   +0.147955      +0.01%        98.34 ms/ev ->       114.05 ms/ev convTrackCandidates
   +0.121636      +0.00%        45.74 ms/ev ->        51.66 ms/ev allConversions
   +0.095129      +0.01%        73.80 ms/ev ->        81.17 ms/ev photonConvTrajSeedFromSingleLeg
   +0.089197      +0.01%        80.69 ms/ev ->        88.22 ms/ev duplicateTrackCandidates
   +0.035194      +0.01%       211.78 ms/ev ->       219.37 ms/ev trackExtrapolator

```

- comparisons in MTV plots are roughly in line with https://mkortela.web.cern.ch/mkortela/tracking/validation/CMSSW_8_1_0_pre5_detachedTriplet/
- Kshort yields clearly increase ( @makortel : your MTV comparison didn't show it, this is probably interplay from the other PRs used as a baseline in my tests; but maybe it's about the quality selections that I didn't apply)
all_sign713-histatvssign710-histat_ttbar13tevpu2017wf10224p0c_recovertexcompositecandidates_generalv0candidates_kshort_reco_obj_mass
all_sign713-histatvssign710-histat_ttbar13tevpu2017wf10224p0c_recovertexcompositecandidates_generalv0candidates_kshort_reco_obj_pt
wf10224_kshort_effs

**Comment by makortel** [comment]
@slava77 Double-checking the DQM root files behind my plots, I can confirm an increase of reconstructed Kshorts by ~30 %. My MTV plots didn't show it because the mass plots are normalized to unit area. For the future I'll add stat boxes to these plots so that a change in the total number of Kshorts can be spotted. Thanks for pointing this out.

**Comment by davidlange6** [comment]
+1

**Pull request merged**